# Mitochondrial depletion in pluripotent stem cells and embryos 

This repository contains the code associated with the paper titled "Mitochondrial depletion in pluripotent stem cells and embryos."

## Data Analysis

The data analysis for this project is divided into three main parts:
The data analysis was primarily conducted using R software. This project mainly includes three parts: transcriptomic data analysis, proteomic data analysis, and phosphoproteomic data analysis.

### Transcriptomic Data Analysis Code
- **Purpose**: Used for quality control, gene counts normalization, differential expression genes (DEGs) calculation, function annotation, and visualization.
- **Contents**: Scripts.

### Proteomic Data Analysis Code
- **Purpose**: Used for quality control, protein counts normalization, differential expression proteins (DEPs) calculation, function annotation, and visualization.
- **Contents**: Scripts.

### Phosphoproteomic Data Analysis Code
- **Purpose**: Used for quality control, phosphopeptide counts normalization, differential expression phosphorylation proteins calculation, function annotation, and visualization.
- **Contents**: Scripts.

### Marker Gene Files
- **Purpose**: Indicate the cell differentiate trends of mitochondrial depletion cells based on DEGs and DEPs
- **Contents**: Files
- **Reference**: CellMarker 2.0; GSEA database

For more detailed information, please refer to the respective scripts and documentation within this repository.
